interfaces/security.go

```diff
@@ -52,16 +54,69 @@ func (aa *appArmor) securitySystem() SecuritySystem {
 	return SecurityAppArmor
 }
 
-func (aa *appArmor) pathForApp(snapName, appName string) string {
-	return fmt.Sprintf("/run/snappy/security/apparmor/%s/%s.profile", snapName, appName)
-}
-
-func (aa *appArmor) headerForApp(snapName, appName string) []byte {
-	// TODO: use a real header here
-	return []byte(fmt.Sprintf("fake \"/snaps/%s/current/%s\" {\n", snapName, appName))
-}
-
-func (aa *appArmor) footerForApp(snapName, appName string) []byte {
+func (aa *appArmor) pathForApp(snapName, snapVersion, snapOrigin, appName string) string {
+	return fmt.Sprintf("/var/lib/snappy/apparmor/profiles/%s.%s_%s_%s",
+		snapName, snapOrigin, appName, snapVersion)
+}
+
+func (aa *appArmor) headerForApp(snapName, snapVersion, snapOrigin, appName string) []byte {
+	header := string(appArmorHeader)
+	vars := aa.varsForApp(snapName, snapVersion, snapOrigin, appName)
+	profileAttach := aa.profileAttachForApp(snapName, snapVersion, snapOrigin, appName)
+	header = strings.Replace(header, "###VAR###\n", vars, 1)
+	header = strings.Replace(header, "###PROFILEATTACH###", profileAttach, 1)
+	// The rest of those are not used so let's just be explicit about it.
+	// They are unused because they are replaced with equivalent content coming
+	// from various snippets (perhaps not organized as neatly but having the
+	// same meaning).
+	header = strings.Replace(header, "###ABSTRACTIONS###", "", 1)
+	header = strings.Replace(header, "###POLICYGROUPS###", "", 1)
+	header = strings.Replace(header, "###READS###", "", 1)
+	header = strings.Replace(header, "###WRITES###", "", 1)
+	return []byte(header)
+}
+
+func (aa *appArmor) varsForApp(snapName, snapVersion, snapOrigin, appName string) string {
+	return fmt.Sprintf(`
+# Specified profile variables
+@{APP_APPNAME}="%s"
+@{APP_ID_DBUS}="%s"
+@{APP_PKGNAME_DBUS}="%s"
+@{APP_PKGNAME}="%s"
+@{APP_VERSION}="%s"
+@{INSTALL_DIR}="{/snaps,/gadget}"
+# Deprecated:
+@{CLICK_DIR}="{/snaps,/gadget}"
+`,
+		appName,
+		dbusPath(fmt.Sprintf("%s.%s_%s_%s", snapName, snapOrigin, appName, snapVersion)),
+		dbusPath(fmt.Sprintf("%s.%s", snapName, snapOrigin)),
+		fmt.Sprintf("%s.%s", snapName, snapOrigin),
+		snapVersion)
+}
+
+func (aa *appArmor) profileAttachForApp(snapName, snapVersion, snapOrigin, appName string) string {
+	return fmt.Sprintf("profile \"%s.%s_%s_%s\"",
+		snapName, snapOrigin, appName, snapVersion)
+}
+
+// Generate a string suitable for use in a DBus object
+func dbusPath(s string) string {
+	const allowed = `abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789`
+	buf := bytes.NewBuffer(make([]byte, 0, len(s)))
+
+	for _, c := range []byte(s) {
+		if strings.IndexByte(allowed, c) >= 0 {
+			fmt.Fprintf(buf, "%c", c)
+		} else {
+			fmt.Fprintf(buf, "_%02x", c)
+		}
+	}
+
+	return buf.String()
+}
+
```

Comment:
dbusPath here is a near code copy of dbusPath in ./snappy/security.go. What are the plans for snappy/security.go? I'm guessing it needs to stick around for old-security/security-policy and old-security/security-override? The former is going to need the same sorts of apparmor variables so shouldn't dbusPath() be refactored? I imagine there is other similar code that could be refactored in snappy/security.go and interfaces/security.go surrounding old-security and interfaces.